## Assignment 2 Pandas

### Note: make sure to save your work and double-check the uploaded file. Empty assignment or no submission receives 0 point, regardless of the reason.

**Instructions:** <br>
a. write your answer after `# your answer here`. Do not delete any code that is in the cell, as the code facilitates checking your answers. If you happen to delete any code that prevents expected outputs, you may lose points. <br>
b. when you finish the assignment, make sure to use `Kernel` -> `Restart and Run All` to restart the kernel and rerun the entire notebook to generate the desired outputs, before submitting your assignment. <br>
c. upload your .ipynb file. <br>
<br>
Each subquestion is worth 5 points, and the total score for assignment 2 is 75 points.

**Policy:** <br>
Assignment submitted late will receive a late penalty (30% point reduction), regardless of the reason. Empty assignment or no submission receives 0 point.


### Question 1

Use the code to simulate data and answer the questions below.

In [1]:
import pandas as pd
import numpy as np
np.random.seed(1234)
s1 = pd.Series(np.random.randint(1, high=5, size=100, dtype='l'))
s2 = pd.Series(np.random.randint(1, high=4, size=100, dtype='l'))
s3 = pd.Series(np.random.randint(100, high=301, size=100, dtype='l'))
s4 = pd.Series(np.random.randint(10000, high=30001, size=100, dtype='l'))

 1. (1) Create a DataFrame `df` by concatenating the three Series s1, s2, s3 as three columns (df: 100 by 3), and then rename column names to be 'a', 'b', 'c'. Keep the first 70 rows of `df`(df: 70 by 3).<br>
 Create another DataFrame `df2` by joinning the two Series s1, s4 by column (df2: 100 by 2), and the column names should be 'A' and 'B'. Then, keep the last 50 rows of `df2`(df2: 50 by 2).

In [9]:
# Your answer here
df = pd.concat([s1, s2, s3], axis=1)
df.columns = ['a', 'b', 'c']
df = df.head(70)

df2 = pd.concat([s1, s4], axis=1)
df2.columns = ['A', 'B']
df2 = df2.tail(50)

# for grading
print(df.columns,'\n'); print(df2.columns,'\n')
print(df.index,'\n'); print(df2.index,'\n')

Index(['a', 'b', 'c'], dtype='object') 

Index(['A', 'B'], dtype='object') 

RangeIndex(start=0, stop=70, step=1) 

RangeIndex(start=50, stop=100, step=1) 



1. (2) Merge `df` and `df2` by the index values into `df3`. Note that we will only keep the intersections of the two dataframes (`df3`: 20 by 5). Additionally, drop the column in `df3` that is repetitive (`df3`: 20 by 4).

In [30]:
# Your answer here
df3 = pd.merge(df, df2, left_index=True, right_index=True).drop('A', axis=1)

# for grading
df3.head()

a  b    c      B
50  2  1  168  19055
51  4  3  160  16122
52  4  1  206  22849
53  4  1  103  16809
54  3  1  275  29924

1. (3) Create a one column DataFrame with the values of the three Series s1, s2, s3 and assign it to 'df4' (300 by 1).

In [39]:
# Your answer here
# print(s1)
# print(s2)
# print(s3)
df4 = pd.concat([s1, s2, s3], axis=0)

# for grading
print(type(df4)) #check the type of the df4
print(df4.shape) #check the shape
print(df4.tail()) #check last columns: note the index

<class 'pandas.core.series.Series'>
(300,)
295    202
296    108
297    141
298    261
299    169
dtype: int64


1. (4) Reindex the DataFrame so that the index is from 0 to 299

In [40]:
# Your answer here
df4 = df4.reset_index(drop=True)

# for grading
df4

0        4
1        4
2        3
3        2
4        1
      ... 
295    202
296    108
297    141
298    261
299    169
Length: 300, dtype: int64

### Question 2

First, download the Olympic medal dataset (medals.csv). Use the following code to load the dataset.

The columns give information on: <br>
summer Olympic games--number of games and summer medals;  <br>
winter Olympic games--number of games and winter medals;  <br>
total Olympic games--number of games and total medals.

In [43]:
import pandas as pd
df = pd.read_csv('medals.csv', index_col=0, skiprows=0)

2. (1) Show the first 6 rows in the DataFrame.

In [44]:
# Your answer here
df.head(n=6)

Summer  Gold  Silver  Bronze  Total   Winter  Gold.1  Silver.1  \
Afghanistan       13     0       0       2      2        0       0         0   
Algeria           12     5       2       8     15        3       0         0   
Argentina         23    18      24      28     70       18       0         0   
Armenia            5     1       2       9     12        6       0         0   
Australasia        2     3       4       5     12        0       0         0   
Australia         25   139     152     177    468       18       5         3   

             Bronze.1  Total.1   Games  Gold.2  Silver.2  Bronze.2  \
Afghanistan         0        0      13       0         0         2   
Algeria             0        0      15       5         2         8   
Argentina           0        0      41      18        24        28   
Armenia             0        0      11       1         2         9   
Australasia         0        0       2       3         4         5   
Australia           4       12      43     144       155       181   

             Combined total   ID  
Afghanistan               2  AFG  
Algeria                  15  ALG  
Argentina                70  ARG  
Armenia                  12  ARM  
Australasia              12  ANZ  
Australia               480  AUS

2. (2) Provide the shape of the DataFrame.

In [45]:
# Your answer here
df.shape

(147, 16)

2. (3) Rename the column name 'Combined total' to 'Combined Total' and then show the columns.

In [47]:
# Your answer here
df.rename(columns={ 'Combined total': 'Combined Total' }, inplace=True)

#for grading
df.columns[14]

'Combined Total'

2. (4) Change all column names to lowercase and remove all spaces ' '.

In [52]:
# Your answer here
df.rename(str.lower, axis='columns', inplace=True)
df.columns = df.columns.str.replace(' ', '')

# for grading
df.columns

Index(['summer', 'gold', 'silver', 'bronze', 'total', 'winter', 'gold.1',
       'silver.1', 'bronze.1', 'total.1', 'games', 'gold.2', 'silver.2',
       'bronze.2', 'combinedtotal', 'id'],
      dtype='object')

2. (5) Which country has won the most gold medals in summer games? Provide the country name and the number of gold medals. <br>
Expected output: a tuple consisting of the country name and the number of gold medals.

In [74]:
def most_gold():
    # Your answer here
    if ('Totals' in df.index):
        df.drop(index="Totals", inplace=True)
        
    most_gold_row = df[df['gold'] == df['gold'].max()]
    
    countryName = most_gold_row.index[0]
    goldNumber = most_gold_row.iloc[0]['gold']
    
    return countryName,goldNumber

most_gold()

('United States', 976)

2. (6) Which country had the largest difference between their summer and winter medal counts, relative to the total medal count? <br> Note: only include countries that have won at least 10 medals in both summer and winter. <br>
$$\frac{Summer_{Metal} - Winter_{Metal}}{Total_{Metal}}$$
Instruction: create a new copy of the DataFrame, first calculate the value as a new column named 'percent', and then **return the row information of the country as a pandas Series**, i.e.:<br> <br>
summer x<br>
gold x<br>
silver x<br>
...<br>
percent x<br>
Name: x, dtype: object




In [88]:
def max_diff():
    # Your answer here
    filtered_df = df[(df["gold"] >= 10) & (df["gold.1"] >= 10)]
    filtered_df = filtered_df.copy()
    filtered_df['percent'] = (abs(filtered_df['gold'] - filtered_df['gold.1'])) / filtered_df['gold.2']
    return filtered_df[filtered_df['percent'] == filtered_df['percent'].max()]

max_diff()

summer  gold  silver  bronze  total  winter  gold.1  silver.1  \
Great Britain      27   236     272     272    780      22      10         4   

               bronze.1  total.1  games  gold.2  silver.2  bronze.2  \
Great Britain        12       26     49     246       276       284   

               combinedtotal   id   percent  
Great Britain            806  GBR  0.918699

2. (7) Create a new column 'country_cat' that denotes the first letter of the country name, i.e. "A" for "Afghanistan", and count the number of countries in each category. Make sure to present the results alphabetically.

In [97]:
def country_first_letter():
    # Your answer here
    df['country_cat'] = df.index.str[0]
    country_counts = df['country_cat'].value_counts().sort_index()
    return country_counts


country_first_letter()

A     8
B    12
C    12
D     3
E     6
F     2
G     9
H     3
I    10
J     2
K     4
L     5
M    10
N     8
P     8
Q     1
R     3
S    17
T     8
U     8
V     3
W     1
Y     1
Z     2
Name: country_cat, dtype: int64

# Polars

Complete the following tasks using Polars code (No Pandas!)</font></h3>

* For (1) the column names you print should only include the selected columns
* Your answer to (3) and (4) must include **both** the statistics **and** the interpretation of the results. Guessing the relationship between age and weight without evidence (code output) to support your interpretation in (4) will not be accepted.
    
3. (1) Read in the data set ```crab-data.csv``` and assign it the name ```df```. Select the columns ```Age```, ```Weight```, ```Height```, and ```Diameter```. Print the column names. <br><br>

In [99]:
pip install polars==0.19.11

  Obtaining dependency information for polars==0.19.11 from https://files.pythonhosted.org/packages/d1/b3/dfa5a51d90da4fcda5c95dcaabd100b5db544e7237a741da3408bddfdb23/polars-0.19.11-cp38-abi3-macosx_11_0_arm64.whl.metadata
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 22.9/22.9 MB 4.7 MB/s eta 0:00:0000:0100:01
Note: you may need to restart the kernel to use updated packages.


In [122]:
import polars as pl
# write your code below
df = pl.read_csv('crab-data.csv')
df = df.select(['Age', 'Weight', 'Height', 'Diameter'])
print(df.columns)

['Age', 'Weight', 'Height', 'Diameter']


3. (2) Consider a subset of the data that only includes crabs with ```Height``` greater than 0.5 **OR** ```Diameter``` greater than 1. How many crabs are there in the subset?  <br> <br>

In [123]:
# write your code below
subset = df.filter((df['Height'] > 0.5) | (df['Diameter'] > 1))
num_crabs = subset.shape[0]
print(num_crabs)

44386


3. (3) Using the **original data frame** ```df``` **from Part 3.(1)**, create a new variable ```age_quartile``` that categorizes the crabs based on which quartile their age belongs to. In other words, ```age_quartile``` should be encoded as follows:
    * **When** age is less than the 25th quantile **then** ```age_quartile = 1```
    * **When** age is greater than or equal to the 25th quantile **and** less than the 50th quantile **then** ```age_quartile = 2```
    * **When** age is greater than or equal to the 50th quantile **and** less than the 75th quantile **then** ```age_quartile = 3```
    * **When** age is greater than or equal to the 75th quantile **then** ```age_quartile = 4```
   
Then calculate the mean ```Weight``` and standard deviation of ```Height``` for the crabs within each category of ```age_quartile```. Based on your results, is the average weight for older crabs higher or lower than the average weight for younger crabs?

In [160]:
# write your code below
q25 = df['Age'].quantile(0.25)
q50 = df['Age'].quantile(0.5)
q75 = df['Age'].quantile(0.75)

df = df.with_columns(
    pl.when(df['Age'] < q25).then(1)
    .when((df['Age'] >= q25) & (df['Age'] < q50)).then(2)
    .when((df['Age'] >= q50) & (df['Age'] < q75)).then(3)
    .when(df['Age'] >= q75).then(4)
    .otherwise(None).alias('age_quartile')
)

df2 = df.group_by(pl.col('age_quartile')).agg(pl.col(['Weight']).mean().alias('Mean Weight')).sort(by='age_quartile')
df3 = df.group_by(pl.col('age_quartile')).agg(pl.col(['Height']).mean().alias('Mean Height')).sort(by='age_quartile')
df4 = df.group_by(pl.col('age_quartile')).agg(pl.col(['Weight']).std().alias('SD Weight')).sort(by='age_quartile')
df5 = df.group_by(pl.col('age_quartile')).agg(pl.col(['Height']).std().alias('SD Height')).sort(by='age_quartile')

df6 = df2.join(df4, on='age_quartile').join(df3, on='age_quartile').join(df5, on='age_quartile')

print(df)
print(df6)

shape: (74_051, 6)
┌─────┬───────────┬────────┬──────────┬─────────┬──────────────┐
│ Age ┆ Weight    ┆ Height ┆ Diameter ┆ literal ┆ age_quartile │
│ --- ┆ ---       ┆ ---    ┆ ---      ┆ ---     ┆ ---          │
│ i64 ┆ f64       ┆ f64    ┆ f64      ┆ i32     ┆ i32          │
╞═════╪═══════════╪════════╪══════════╪═════════╪══════════════╡
│ 9   ┆ 28.973189 ┆ 0.375  ┆ 1.175    ┆ 1       ┆ 2            │
│ 8   ┆ 10.418441 ┆ 0.275  ┆ 0.825    ┆ 1       ┆ 2            │
│ 9   ┆ 24.777463 ┆ 0.375  ┆ 1.1125   ┆ 1       ┆ 2            │
│ 11  ┆ 50.660556 ┆ 0.5    ┆ 1.4125   ┆ 1       ┆ 4            │
│ …   ┆ …         ┆ …      ┆ …        ┆ …       ┆ …            │
│ 6   ┆ 10.446791 ┆ 0.275  ┆ 0.8625   ┆ 1       ┆ 1            │
│ 10  ┆ 29.48348  ┆ 0.4125 ┆ 1.2      ┆ 1       ┆ 3            │
│ 8   ┆ 16.768729 ┆ 0.3125 ┆ 0.9625   ┆ 1       ┆ 2            │
│ 6   ┆ 5.386405  ┆ 0.2    ┆ 0.675    ┆ 1       ┆ 1            │
└─────┴───────────┴────────┴──────────┴─────────┴──────────────┘
shape:

3. (4) Write your interpretation for Part 3 (3).

**Interpretation:**
We can observe that the average weight increases as ```age_quartile``` value increases. This suggests younger crabs have higher average weight than older crabs. The average weight tends to increase as the crabs get older. It shows a positive relationship between age and weight of the crabs.